LibreNMS/Data/Store/Kafka.php

```diff
@@ -0,0 +1,304 @@
+            }
+        }
+
+        return new Producer($conf);
+    }
+
+    public function safeFlush()
+    {
+        // check if client instance exists
+        if ($this->client === null) {
+            return;
+        }
+
+        try {
+            // get total number of messages in the queue
+            $outQLen = $this->client->getOutQLen();
+
+            if ($outQLen > 0) {
+                // During shutdown, Log facades might not work properly, use d_echo as fallback
+                error_log("KAFKA: SafeFlush | Flushing {$outQLen} remaining messages");
+
+                // Use cached timeout value to avoid Config during shutdown
+                $result = $this->client->flush($this->kafkaFlushTimeout);
+
+                if (RD_KAFKA_RESP_ERR_NO_ERROR !== $result) {
+                    $error_msg = sprintf(
+                        'KAFKA: SafeFlush | Flush failed. Error: %s, Code: %d, Device ID: %d, Remaining: %d',
+                        Library::rd_kafka_err2str($result),
+                        $result,
+                        $this->device_id,
+                        $this->client->getOutQLen()
+                    );
+
+                    error_log($error_msg);
+                }
+            }
+        } catch (\Throwable $e) {
+            $error_msg = 'KAFKA: SafeFlush | failed with exception. Error: ' . $e->getMessage() . '. Trace: ' . $e->getTraceAsString();
+            error_log($error_msg);
+        } finally {
+            // Reset the client to null to avoid further operations
+            $this->client = null;
+        }
+    }
+
+    public function getName(): string
+    {
+        return 'Kafka';
+    }
+
+    public static function isEnabled(): bool
+    {
+        return Config::get('kafka.enable', false);
+    }
+
+    public function getKafkaFlushTimeout()
+    {
+        return $this->kafkaFlushTimeout;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function write(string $measurement, array $fields, array $tags = [], array $meta = []): void
+    {
+        try {
+            $device = $this->getDevice($meta);
+            // get the singleton instance of the produced
+            /** @var Producer $producer */
+            $producer = $this->client;
+            $this->device_id = $device['device_id'];
+            $topic = $producer->newTopic(Kafka::getTopicName());
+
+            $device_data = DeviceCache::get($device['device_id']);
+            $excluded_groups = [];
+            $excluded_measurement = [];
+
+            // Load excluded values from config
+            foreach (Config::get('kafka.groups-exclude', []) as $exclude_group) {
+                $excluded_groups[] = strtoupper($exclude_group);
+            }
+            foreach (Config::get('kafka.measurement-exclude', []) as $exclude_measurement) {
+                $excluded_measurement[] = strtoupper($exclude_measurement);
+            }
```

Comment:
These should be pre-calculated and stored as properties, they should not change during a run.